File: concepts/08-security/node/attack-vectors.js
/**
 * Common Attack Vectors
 * Demonstrates offensive security concepts - attack detection and prevention
 */

/**
 * Attack Detector
 * Detects common attack patterns
 */
class AttackDetector {
  /**
   * Detect SQL Injection
   */
  static detectSQLInjection(input) {
    const patterns = [
      /(\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE)\b)/i,
      /(--|#|\/\*|\*\/)/,
      /(;|\||&)/,
      /(UNION\s+SELECT)/i,
      /(OR\s+1\s*=\s*1)/i,
      /('|"|`)/,
      /(\bxp_\w+)/i,
      /(WAITFOR\s+DELAY)/i
    ];
    
    for (const pattern of patterns) {
      if (pattern.test(input)) {
        return {
          detected: true,
          attackType: 'SQL Injection',
          pattern: pattern.toString(),
          severity: 'HIGH',
          input: input.substring(0, 100)
        };
      }
    }
    
    return { detected: false };
  }

  /**
   * Detect XSS
   */
  static detectXSS(input) {
    const patterns = [
      /<script[^>]*>.*?<\/script>/gi,
      /<iframe[^>]*>.*?<\/iframe>/gi,
      /javascript:/gi,
      /on\w+\s*=/gi,
      /<img[^>]+src[^>]*=.*javascript:/gi,
      /<svg[^>]*onload/gi,
      /<body[^>]*onload/gi,
      /<input[^>]*onfocus/gi
    ];
    
    for (const pattern of patterns) {
      if (pattern.test(input)) {
        return {
          detected: true,
          attackType: 'Cross-Site Scripting (XSS)',
          pattern: pattern.toString(),
          severity: 'HIGH',
          input: input.substring(0, 100)
        };
      }
    }
    
    return { detected: false };
  }

  /**
   * Detect Command Injection
   */
  static detectCommandInjection(input) {
    const patterns = [
      /[;&|`$(){}[\]]/,
      /(\||\|\|)/,
      /(&&|&)/,
      /(;|;)/,
      /(`|`)/,
      /(\$\(|`)/,
      /(>\s*\/dev\/null)/,
      /(<\s*\/dev\/null)/
    ];
    
    for (const pattern of patterns) {
      if (pattern.test(input)) {
        return {
          detected: true,
          attackType: 'Command Injection',
          pattern: pattern.toString(),
          severity: 'CRITICAL',
          input: input.substring(0, 100)
        };
      }
    }
    
    return { detected: false };
  }

  /**
   * Detect Path Traversal
   */
  static detectPathTraversal(input) {
    const patterns = [
      /\.\./,
      /\.\.\/\.\./,
      /\.\.\\/,
      /\/\.\./,
      /\.\.%2f/i,
      /\.\.%5c/i,
      /\.\.%252f/i,
      /\.\.%255c/i
    ];
    
    for (const pattern of patterns) {
      if (pattern.test(input)) {
        return {
          detected: true,
          attackType: 'Path Traversal',
          pattern: pattern.toString(),
          severity: 'HIGH',
          input: input.substring(0, 100)
        };
      }
    }
    
    return { detected: false };
  }

  /**
   * Detect LDAP Injection
   */
  static detectLDAPInjection(input) {
    const patterns = [
      /[()&|!]/,
      /(\*|\(|\)|&|\|)/,
      /(cn=|ou=|dc=)/i
    ];
    
    for (const pattern of patterns) {
      if (pattern.test(input)) {
        return {
          detected: true,
          attackType: 'LDAP Injection',
          pattern: pattern.toString(),
          severity: 'MEDIUM',
          input: input.substring(0, 100)
        };
      }
    }
    
    return { detected: false };
  }

  /**
   * Detect XML Injection
   */
  static detectXMLInjection(input) {
    const patterns = [
      /<!\[CDATA\[/,
      /<!ENTITY/,
      /<\!DOCTYPE/,
      /<script/gi,
      /<iframe/gi
    ];
    
    for (const pattern of patterns) {
      if (pattern.test(input)) {
        return {
          detected: true,
          attackType: 'XML Injection',
          pattern: pattern.toString(),
          severity: 'HIGH',
          input: input.substring(0, 100)
        };
      }
    }
    
    return { detected: false };
  }

  /**
   * Detect Brute Force Attempt
   */
  static detectBruteForce(attempts, timeWindow = 300000) {
    const recentAttempts = attempts.filter(
      attempt => Date.now() - attempt.timestamp < timeWindow
    );
    
    if (recentAttempts.length > 10) {
      return {
        detected: true,
        attackType: 'Brute Force',
        attempts: recentAttempts.length,
        severity: 'MEDIUM',
        recommendation: 'Implement account lockout'
      };
    }
    
    return { detected: false };
  }

  /**
   * Comprehensive attack detection
   */
  static detectAll(input) {
    const detections = [];
    
    const checks = [
      this.detectSQLInjection,
      this.detectXSS,
      this.detectCommandInjection,
      this.detectPathTraversal,
      this.detectLDAPInjection,
      this.detectXMLInjection
    ];
    
    for (const check of checks) {
      const result = check(input);
      if (result.detected) {
        detections.push(result);
      }
    }
    
    return {
      detected: detections.length > 0,
      attacks: detections,
      count: detections.length
    };
  }
}

/**
 * Security Scanner
 * Simulates vulnerability scanning
 */
class SecurityScanner {
  constructor() {
    this.vulnerabilities = [];
  }

  /**
   * Scan for SQL injection vulnerabilities
   */
  scanSQLInjection(endpoints) {
    const vulnerable = [];
    
    for (const endpoint of endpoints) {
      // Simulate testing
      const testInputs = [
        "admin' OR '1'='1",
        "'; DROP TABLE users--",
        "1' UNION SELECT * FROM users--"
      ];
      
      for (const testInput of testInputs) {
        const detection = AttackDetector.detectSQLInjection(testInput);
        if (detection.detected) {
          vulnerable.push({
            endpoint,
            vulnerability: 'SQL Injection',
            severity: 'HIGH',
            testInput: testInput.substring(0, 50)
          });
        }
      }
    }
    
    return vulnerable;
  }

  /**
   * Scan for XSS vulnerabilities
   */
  scanXSS(endpoints) {
    const vulnerable = [];
    
    for (const endpoint of endpoints) {
      const testInputs = [
        '<script>alert("XSS")</script>',
        '<img src=x onerror=alert("XSS")>',
        '<svg onload=alert("XSS")>'
      ];
      
      for (const testInput of testInputs) {
        const detection = AttackDetector.detectXSS(testInput);
        if (detection.detected) {
          vulnerable.push({
            endpoint,
            vulnerability: 'Cross-Site Scripting',
            severity: 'HIGH',
            testInput: testInput.substring(0, 50)
          });
        }
      }
    }
    
    return vulnerable;
  }

  /**
   * Comprehensive scan
   */
  scan(endpoints) {
    const results = {
      sqlInjection: this.scanSQLInjection(endpoints),
      xss: this.scanXSS(endpoints),
      timestamp: Date.now()
    };
    
    const totalVulnerabilities = results.sqlInjection.length + results.xss.length;
    
    return {
      ...results,
      totalVulnerabilities,
      riskLevel: totalVulnerabilities > 5 ? 'HIGH' : 
                 totalVulnerabilities > 2 ? 'MEDIUM' : 'LOW'
    };
  }
}

/**
 * Threat Model
 */
class ThreatModel {
  constructor() {
    this.threats = [];
  }

  /**
   * Add threat using STRIDE model
   */
  addThreat(component, threatType, description, mitigation) {
    this.threats.push({
      component,
      threatType, // Spoofing, Tampering, Repudiation, Information Disclosure, DoS, Elevation
      description,
      mitigation,
      severity: this.calculateSeverity(threatType),
      timestamp: Date.now()
    });
  }

  /**
   * Calculate severity using DREAD
   */
  calculateSeverity(threatType) {
    const severityMap = {
      'Spoofing': 'MEDIUM',
      'Tampering': 'HIGH',
      'Repudiation': 'LOW',
      'Information Disclosure': 'HIGH',
      'DoS': 'MEDIUM',
      'Elevation': 'CRITICAL'
    };
    
    return severityMap[threatType] || 'MEDIUM';
  }

  /**
   * Get threats by component
   */
  getThreatsByComponent(component) {
    return this.threats.filter(t => t.component === component);
  }

  /**
   * Get all threats
   */
  getAllThreats() {
    return this.threats;
  }
}

// Example usage
function demonstrateAttackVectors() {
  console.log('=== Attack Vector Detection ===\n');
  
  // Test various attack inputs
  const testInputs = [
    "admin' OR '1'='1",
    '<script>alert("XSS")</script>',
    '../../etc/passwd',
    '; rm -rf /',
    'normal input'
  ];
  
  for (const input of testInputs) {
    const detection = AttackDetector.detectAll(input);
    if (detection.detected) {
      console.log(`Input: ${input.substring(0, 30)}...`);
      console.log('Detected attacks:', detection.attacks);
      console.log('');
    }
  }
  
  console.log('=== Security Scanning ===\n');
  const scanner = new SecurityScanner();
  const endpoints = ['/api/users', '/api/products', '/api/orders'];
  const scanResults = scanner.scan(endpoints);
  console.log('Scan results:', scanResults);
  
  console.log('\n=== Threat Modeling ===\n');
  const threatModel = new ThreatModel();
  threatModel.addThreat(
    'Authentication Service',
    'Spoofing',
    'Attacker may impersonate users',
    'Implement MFA and strong password policies'
  );
  threatModel.addThreat(
    'Database',
    'Information Disclosure',
    'SQL injection may expose sensitive data',
    'Use parameterized queries and input validation'
  );
  threatModel.addThreat(
    'API Gateway',
    'DoS',
    'DDoS attacks may overwhelm the system',
    'Implement rate limiting and DDoS protection'
  );
  
  console.log('Threats identified:', threatModel.getAllThreats());
}

if (require.main === module) {
  demonstrateAttackVectors();
}

module.exports = { AttackDetector, SecurityScanner, ThreatModel };

File: concepts/08-security/python/attack_vectors.py
"""
Common Attack Vectors
Demonstrates offensive security concepts - attack detection and prevention
"""
import re
from typing import List, Dict, Any, Optional
from dataclasses import dataclass, field
from datetime import datetime


@dataclass
class AttackDetection:
    """Attack detection result"""
    detected: bool
    attack_type: Optional[str] = None
    pattern: Optional[str] = None
    severity: Optional[str] = None
    input: Optional[str] = None


class AttackDetector:
    """Attack detector implementation"""
    
    @staticmethod
    def detect_sql_injection(input_str: str) -> Dict[str, Any]:
        """Detect SQL Injection"""
        patterns = [
            re.compile(r'\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE)\b', re.IGNORECASE),
            re.compile(r'(--|#|/\*|\*/)'),
            re.compile(r'(;|\||&)'),
            re.compile(r'UNION\s+SELECT', re.IGNORECASE),
            re.compile(r'OR\s+1\s*=\s*1', re.IGNORECASE),
            re.compile(r'[\'"`]'),
            re.compile(r'\bxp_\w+', re.IGNORECASE),
            re.compile(r'WAITFOR\s+DELAY', re.IGNORECASE)
        ]
        
        for pattern in patterns:
            if pattern.search(input_str):
                return {
                    'detected': True,
                    'attack_type': 'SQL Injection',
                    'pattern': pattern.pattern,
                    'severity': 'HIGH',
                    'input': input_str[:100]
                }
        
        return {'detected': False}
    
    @staticmethod
    def detect_xss(input_str: str) -> Dict[str, Any]:
        """Detect XSS"""
        patterns = [
            re.compile(r'<script[^>]*>.*?</script>', re.IGNORECASE | re.DOTALL),
            re.compile(r'<iframe[^>]*>.*?</iframe>', re.IGNORECASE | re.DOTALL),
            re.compile(r'javascript:', re.IGNORECASE),
            re.compile(r'on\w+\s*=', re.IGNORECASE),
            re.compile(r'<img[^>]+src[^>]*=.*javascript:', re.IGNORECASE),
            re.compile(r'<svg[^>]*onload', re.IGNORECASE),
            re.compile(r'<body[^>]*onload', re.IGNORECASE),
            re.compile(r'<input[^>]*onfocus', re.IGNORECASE)
        ]
        
        for pattern in patterns:
            if pattern.search(input_str):
                return {
                    'detected': True,
                    'attack_type': 'Cross-Site Scripting (XSS)',
                    'pattern': pattern.pattern,
                    'severity': 'HIGH',
                    'input': input_str[:100]
                }
        
        return {'detected': False}
    
    @staticmethod
    def detect_command_injection(input_str: str) -> Dict[str, Any]:
        """Detect Command Injection"""
        patterns = [
            re.compile(r'[;&|`$(){}[\]]'),
            re.compile(r'(\||\|\|)'),
            re.compile(r'(&&|&)'),
            re.compile(r'(;|;)'),
            re.compile(r'(`|`)'),
            re.compile(r'(\$\(|`)'),
            re.compile(r'>\s*/dev/null'),
            re.compile(r'<\s*/dev/null')
        ]
        
        for pattern in patterns:
            if pattern.search(input_str):
                return {
                    'detected': True,
                    'attack_type': 'Command Injection',
                    'pattern': pattern.pattern,
                    'severity': 'CRITICAL',
                    'input': input_str[:100]
                }
        
        return {'detected': False}
    
    @staticmethod
    def detect_path_traversal(input_str: str) -> Dict[str, Any]:
        """Detect Path Traversal"""
        patterns = [
            re.compile(r'\.\.'),
            re.compile(r'\.\./\.\.'),
            re.compile(r'\.\.\\'),
            re.compile(r'/\.\.'),
            re.compile(r'\.\.%2f', re.IGNORECASE),
            re.compile(r'\.\.%5c', re.IGNORECASE),
            re.compile(r'\.\.%252f', re.IGNORECASE),
            re.compile(r'\.\.%255c', re.IGNORECASE)
        ]
        
        for pattern in patterns:
            if pattern.search(input_str):
                return {
                    'detected': True,
                    'attack_type': 'Path Traversal',
                    'pattern': pattern.pattern,
                    'severity': 'HIGH',
                    'input': input_str[:100]
                }
        
        return {'detected': False}
    
    @staticmethod
    def detect_ldap_injection(input_str: str) -> Dict[str, Any]:
        """Detect LDAP Injection"""
        patterns = [
            re.compile(r'[()&|!]'),
            re.compile(r'(\*|\(|\)|&|\|)'),
            re.compile(r'(cn=|ou=|dc=)', re.IGNORECASE)
        ]
        
        for pattern in patterns:
            if pattern.search(input_str):
                return {
                    'detected': True,
                    'attack_type': 'LDAP Injection',
                    'pattern': pattern.pattern,
                    'severity': 'MEDIUM',
                    'input': input_str[:100]
                }
        
        return {'detected': False}
    
    @staticmethod
    def detect_xml_injection(input_str: str) -> Dict[str, Any]:
        """Detect XML Injection"""
        patterns = [
            re.compile(r'<!\[CDATA\['),
            re.compile(r'<!ENTITY'),
            re.compile(r'<!DOCTYPE'),
            re.compile(r'<script', re.IGNORECASE),
            re.compile(r'<iframe', re.IGNORECASE)
        ]
        
        for pattern in patterns:
            if pattern.search(input_str):
                return {
                    'detected': True,
                    'attack_type': 'XML Injection',
                    'pattern': pattern.pattern,
                    'severity': 'HIGH',
                    'input': input_str[:100]
                }
        
        return {'detected': False}
    
    @staticmethod
    def detect_brute_force(attempts: List[Dict[str, Any]], time_window: float = 300.0) -> Dict[str, Any]:
        """Detect Brute Force Attempt"""
        now = datetime.now().timestamp()
        recent_attempts = [
            attempt for attempt in attempts
            if now - attempt['timestamp'] < time_window
        ]
        
        if len(recent_attempts) > 10:
            return {
                'detected': True,
                'attack_type': 'Brute Force',
                'attempts': len(recent_attempts),
                'severity': 'MEDIUM',
                'recommendation': 'Implement account lockout'
            }
        
        return {'detected': False}
    
    @staticmethod
    def detect_all(input_str: str) -> Dict[str, Any]:
        """Comprehensive attack detection"""
        detections = []
        
        checks = [
            AttackDetector.detect_sql_injection,
            AttackDetector.detect_xss,
            AttackDetector.detect_command_injection,
            AttackDetector.detect_path_traversal,
            AttackDetector.detect_ldap_injection,
            AttackDetector.detect_xml_injection
        ]
        
        for check in checks:
            result = check(input_str)
            if result['detected']:
                detections.append(result)
        
        return {
            'detected': len(detections) > 0,
            'attacks': detections,
            'count': len(detections)
        }


class SecurityScanner:
    """Security scanner implementation"""
    
    def __init__(self):
        self.vulnerabilities: List[Dict[str, Any]] = []
    
    def scan_sql_injection(self, endpoints: List[str]) -> List[Dict[str, Any]]:
        """Scan for SQL injection vulnerabilities"""
        vulnerable = []
        
        for endpoint in endpoints:
            test_inputs = [
                "admin' OR '1'='1",
                "'; DROP TABLE users--",
                "1' UNION SELECT * FROM users--"
            ]
            
            for test_input in test_inputs:
                detection = AttackDetector.detect_sql_injection(test_input)
                if detection['detected']:
                    vulnerable.append({
                        'endpoint': endpoint,
                        'vulnerability': 'SQL Injection',
                        'severity': 'HIGH',
                        'test_input': test_input[:50]
                    })
        
        return vulnerable
    
    def scan_xss(self, endpoints: List[str]) -> List[Dict[str, Any]]:
        """Scan for XSS vulnerabilities"""
        vulnerable = []
        
        for endpoint in endpoints:
            test_inputs = [
                '<script>alert("XSS")</script>',
                '<img src=x onerror=alert("XSS")>',
                '<svg onload=alert("XSS")>'
            ]
            
            for test_input in test_inputs:
                detection = AttackDetector.detect_xss(test_input)
                if detection['detected']:
                    vulnerable.append({
                        'endpoint': endpoint,
                        'vulnerability': 'Cross-Site Scripting',
                        'severity': 'HIGH',
                        'test_input': test_input[:50]
                    })
        
        return vulnerable
    
    def scan(self, endpoints: List[str]) -> Dict[str, Any]:
        """Comprehensive scan"""
        results = {
            'sql_injection': self.scan_sql_injection(endpoints),
            'xss': self.scan_xss(endpoints),
            'timestamp': datetime.now().timestamp()
        }
        
        total_vulnerabilities = len(results['sql_injection']) + len(results['xss'])
        
        return {
            **results,
            'total_vulnerabilities': total_vulnerabilities,
            'risk_level': 'HIGH' if total_vulnerabilities > 5 else
                          'MEDIUM' if total_vulnerabilities > 2 else 'LOW'
        }


class ThreatModel:
    """Threat model implementation"""
    
    def __init__(self):
        self.threats: List[Dict[str, Any]] = []
    
    def add_threat(self, component: str, threat_type: str, description: str, mitigation: str):
        """Add threat using STRIDE model"""
        self.threats.append({
            'component': component,
            'threat_type': threat_type,  # Spoofing, Tampering, Repudiation, Information Disclosure, DoS, Elevation
            'description': description,
            'mitigation': mitigation,
            'severity': self._calculate_severity(threat_type),
            'timestamp': datetime.now().timestamp()
        })
    
    def _calculate_severity(self, threat_type: str) -> str:
        """Calculate severity using DREAD"""
        severity_map = {
            'Spoofing': 'MEDIUM',
            'Tampering': 'HIGH',
            'Repudiation': 'LOW',
            'Information Disclosure': 'HIGH',
            'DoS': 'MEDIUM',
            'Elevation': 'CRITICAL'
        }
        
        return severity_map.get(threat_type, 'MEDIUM')
    
    def get_threats_by_component(self, component: str) -> List[Dict[str, Any]]:
        """Get threats by component"""
        return [t for t in self.threats if t['component'] == component]
    
    def get_all_threats(self) -> List[Dict[str, Any]]:
        """Get all threats"""
        return self.threats


def demonstrate_attack_vectors():
    """Demonstrate attack vectors"""
    print('=== Attack Vector Detection ===\n')
    
    # Test various attack inputs
    test_inputs = [
        "admin' OR '1'='1",
        '<script>alert("XSS")</script>',
        '../../etc/passwd',
        '; rm -rf /',
        'normal input'
    ]
    
    for input_str in test_inputs:
        detection = AttackDetector.detect_all(input_str)
        if detection['detected']:
            print(f"Input: {input_str[:30]}...")
            print('Detected attacks:', detection['attacks'])
            print()
    
    print('=== Security Scanning ===\n')
    scanner = SecurityScanner()
    endpoints = ['/api/users', '/api/products', '/api/orders']
    scan_results = scanner.scan(endpoints)
    print('Scan results:', scan_results)
    
    print('\n=== Threat Modeling ===\n')
    threat_model = ThreatModel()
    threat_model.add_threat(
        'Authentication Service',
        'Spoofing',
        'Attacker may impersonate users',
        'Implement MFA and strong password policies'
    )
    threat_model.add_threat(
        'Database',
        'Information Disclosure',
        'SQL injection may expose sensitive data',
        'Use parameterized queries and input validation'
    )
    threat_model.add_threat(
        'API Gateway',
        'DoS',
        'DDoS attacks may overwhelm the system',
        'Implement rate limiting and DDoS protection'
    )
    
    print('Threats identified:', threat_model.get_all_threats())


if __name__ == '__main__':
    demonstrate_attack_vectors()

